pathfinderlib/main.py

```diff
@@ -25,33 +63,212 @@ def make_fastq_to_fasta_command(self, filename):
+
+    def _load_kegg2ko(self):
+        "Loads the index file and counts how many KOs are assigned to each protein."
+        kegg2ko = pandas.read_csv(self.kegg_to_ko_fp, sep='\t', header=0) # Headers are Subject, KO
+        return kegg2ko.merge(self._getCount(kegg2ko, 'Subject', 'Count_ko'), on='Subject')
+
+    def _parseResults(self, alignment_fp):
+        "Parses an alignment result file. Returns only the columns of interest"
+        colNames = ['# Fields: Query', 'Subject', 'identity', 'e-value']
+        if self.search_method.lower() == "blastx":
+            return pandas.read_csv(alignment_fp, sep='\t', header=None, names=colNames, usecols=[0,1,2,10])
+        else: # for rapsearch
+            return pandas.read_csv(alignment_fp, sep='\t', skiprows=4, usecols=colNames)
+    
+    def _getBestHit(self, alignment):
+        "Finds the best match that passes an e-value threshold."
+        idx = alignment.groupby('# Fields: Query').apply(lambda g: g['e-value'].idxmin())
+        bestAlignment = alignment.ix[idx]
+        return bestAlignment[bestAlignment['e-value']<self.evalue_cutoff]
+    
+    def _map2ko(self, alignment, kegg2ko):
+        """Merges the alignment results with the kegg index file,
+        normalizes the abundances for proteins with multiple KO assignments,
+        finds the total abundances for each ko.
+        """
+        hitCounts = kegg2ko.merge(alignment, on='Subject')
+        hitCounts['norm_abundance'] = hitCounts['Count_hit'] / hitCounts['Count_ko']
+        return self._sumAbundance(hitCounts, 'KO', 'norm_abundance', 'ko_abundance'), hitCounts['Subject'].nunique()
+    
+    def _assign_ko(self, alignment_fp):
+        "Manipulates the alignment file to calculate KO abundances in a fastq file"
+        kegg2ko = self._load_kegg2ko()
+        
+        alignment = self._parseResults(alignment_fp)
+        bestAlignment = self._getBestHit(alignment)
+        numHits = self._getCount(bestAlignment, 'Subject', 'Count_hit')
+
+        # merge and count
+        ko_count, ko_hits = self._map2ko(numHits, kegg2ko)
+        summary = {'mapped_sequences':alignment['# Fields: Query'].nunique(),
+                   'mapped_sequences_evalue':bestAlignment['# Fields: Query'].nunique(),
+                   'unique_prot_hits':bestAlignment['Subject'].nunique(),
+                   'ko_hits':ko_hits,
+                   'unique_ko_hits':ko_count['KO'].nunique()}
+        return ko_count, summary
+
+    def _assign_pathway(self, ko_assign):
+        "Manipulates the KO abundance table to calculate pathway abundances in afastq file"
+        raise NotImplementedError("Pathway abundance calculations are not yet implemented.")
+
+    def run(self, alignment_fp, out_dir):
+        ko_count, summary = self._assign_ko(alignment_fp)
+        ko_count.to_csv(os.path.join(out_dir, os.path.basename(os.path.splitext(alignment_fp)[0]+'.ko')),
+                        sep='\t', index=False)
```

Comment:
Would probably do `ko_fp = os.path.join(out_dir, os.path.basename(os.path.splitext(alignment_fp)[0]+'.ko'))` on the previus line rather than inside the function args.